### PR TITLE
Virtual Machine

### PR DESCRIPTION
**Pull request by janosdebugs**
This pull request adds structures and methods for interactions related to virtual machines.

It is split into separate commits and only adds the base structure as well as NIC creation. Other parts, such as adding disks, etc. are to follow.

### CODE REVIEW

File: client_cluster.go [hunk 1]
```diff
@@ -33,12 +33,15 @@ func convertSDKCluster(sdkCluster *ovirtsdk4.Cluster) (Cluster, error) {
...
-		id:   id,
-		name: name,
+		client: client,
+		id:     id,
+		name:   name,
 	}, nil
 }
 
 type cluster struct {
+	client Client
```
Comment (Gal-Zaidman) [inline client_cluster.go:43]: I don't have anything against adding the client, I even think we talked about something similar when we started this project.
But I think it's safe to say that this is mostly for internal usage, so it will be possible to call other functions that the client provides within the code, because the user that uses the library already have it.
So because it is added to basically all the structs in the library, is it something we can set globally within in library so it will not be a part of each struct ?

File: client_datacenter.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+	id   string
+	name string
+}
+
+func (d datacenter) Clusters(retries ...RetryStrategy) ([]Cluster, error) {
+	return d.client.ListDatacenterClusters(d.id, retries...)
+}
+
+func (d datacenter) HasCluster(clusterID string, retries ...RetryStrategy) (bool, error) {
```
Comment (Gal-Zaidman) [inline client_datacenter.go:53]: This is just an example, but I wonder now that we added functions for the object(datacenter here) how are we distinguishing between a function on the client(i.e DatacenterClient) and a function on the object?

File: client_datacenter_list_clusters.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+			result = make([]Cluster, len(sdkObjects.Slice()))
+			for i, sdkObject := range sdkObjects.Slice() {
+				result[i], e = convertSDKCluster(sdkObject, o)
+				if e != nil {
+					return wrap(e, EBug, "failed to convert cluster during listing item #%d", i)
+				}
+			}
+			return nil
+		})
+	return
```
Comment (Gal-Zaidman) [inline client_datacenter_list_clusters.go:38]: Can you please export it into a separate function like we usually do

File: client_cluster.go (hunk 1 above)
Comment (janosdebugs) [reply client_cluster.go:43]: The `cluster` struct (notice the lower case) implements the `Cluster` interface, so calling parties will never have access to the `client` field. The client is added to the response struct so that we can implement convenience functions on the returned objects, such as getting the list of NIC's directly from the VM object instead of having to go to the client.

File: client_datacenter.go (hunk 2 above)
Comment (janosdebugs) [reply client_datacenter.go:53]: We are not. The functions on the response object are convenience functions that call through to the actual client.

File: client_datacenter_list_clusters.go (hunk 3 above)
Comment (janosdebugs) [reply client_datacenter_list_clusters.go:38]: In this case it's difficult because the function on retry actually uses an input parameter. The only way to do it would be as follows:

```go
func (o *oVirtClient) ListDatacenterClusters(id string, retries ...RetryStrategy) (result []Cluster, err error) {
	retries = defaultRetries(retries, defaultReadTimeouts())
	result = []Cluster{}
	err = retry(
		fmt.Sprintf("listing datacenters %s clusters", id),
		o.logger,
		retries,
                o.doListDatanceterClusters(id),
        )
	return
}

func (o *ovirtClient) doListDatacenterClusters(id string) func() error {
    return func() error {
		response, e := o.conn.
			SystemService().
			DataCentersService().
			DataCenterService(id).
			ClustersService().
			List().
			Send()
		if e != nil {
			return e
		}
		sdkObjects, ok := response.Clusters()
		if !ok {
			return nil
		}
		result = make([]Cluster, len(sdkObjects.Slice()))
		for i, sdkObject := range sdkObjects.Slice() {
			result[i], e = convertSDKCluster(sdkObject, o)
			if e != nil {
				return wrap(e, EBug, "failed to convert cluster during listing item #%d", i)
			}
		}
		return nil
	}
}
```

File: client_cluster.go (hunk 1 above)
Comment (Gal-Zaidman) [reply client_cluster.go:43]: Yep, you are right we are returning the interface so it's ok.
Regarding the second note, we will need to add it on each struct basically, since everything is on the same package wouldn't it be nicer to define it on the global scope? (no strong feeling here just thinking)

Comment (Gal-Zaidman) [reply client_cluster.go:43]: Although on a second thought a user can have multiple clients defined on his program, and if we want to support that use case we can't define it globally

Comment (janosdebugs) [reply client_cluster.go:43]: Exactly, especially in a testing situation having a mock and a live client live next to each other is important. Hidden global scope is evil. :D

File: client_datacenter.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+	GetDatacenter(id string, retries ...RetryStrategy) (Datacenter, error)
+	ListDatacenters(retries ...RetryStrategy) ([]Datacenter, error)
+	ListDatacenterClusters(id string, retries ...RetryStrategy) ([]Cluster, error)
+}
+
+type Datacenter interface {
+	ID() string
+	Name() string
+
+	// Clusters lists the clusters for this datacenter. This is a network call and may be slow.
```
[the remark this answers is not in the document]
Comment (janosdebugs) [reply client_datacenter.go:19]: We can standardize the message, but I feel like there should be something in the naming of these functions to indicate that it's a network call.

File: client_datacenter_list_clusters.go (hunk 3 above)
Comment (Gal-Zaidman) [reply client_datacenter_list_clusters.go:38]: well, it still looks better, maybe call it `listDatacenterClustersInternal` or something similar so it will be clear...
or just define it inside `ListDatacenterClusters` at least that way it is not anonymous

File: client_network.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
+package ovirtclient
+
+import (
+	ovirtsdk4 "github.com/ovirt/go-ovirt"
+)
+
+//go:generate go run scripts/rest.go -i "Network" -n "network"
```
Comment (Gal-Zaidman) [inline client_network.go:7]: I think you forgot the  -o "Network"

Comment (janosdebugs) [reply client_network.go:7]: No, `-o` defaults to the same value as `-i`.

File: client_vm_create.go [hunk 6]
```diff
@@ -1,16 +1,62 @@
...
-func (o *oVirtClient) CreateVM(
-	ctx context.Context,
-	clusterID string,
-	cpuTopo VMCPUTopo,
-	templateID string,
-	blockDevices []VMBlockDevice,
-) {
-	// TODO implement create VM
-	panic("implement me")
+func (o *oVirtClient) CreateVM(clusterID string, name string, templateID string, retries ...RetryStrategy) (result VM, err error) {
```
Comment (Gal-Zaidman) [inline client_vm_create.go]: Since the creation parameters will grow and eventually will be huge maybe we should have a stuct to contain all the creation parameters?

Comment (janosdebugs) [reply client_vm_create.go]: You are right, this is addressed in a later commit.

File: client_vm_create.go [hunk 7]
```diff
@@ -1,16 +1,62 @@
...
+	if err := validateVMCreationParameters(clusterID, name, templateID); err != nil {
+		return nil, err
+	}
+
+	retries = defaultRetries(retries, defaultWriteTimeouts())
+	err = retry(
+		fmt.Sprintf("creating VM %s", name),
+		o.logger,
+		retries,
+		func() error {
```
Comment (Gal-Zaidman) [inline client_vm_create.go:19]: I know we depend on the wrapping function parameters, but we have to export it into a separate function because although now it is small, eventually it will become huge and having such a huge anonymous function is not fun :)

File: client_vm_remove.go [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+	"fmt"
+)
+
+func (o *oVirtClient) RemoveVM(id string, retries ...RetryStrategy) (err error) {
+	retries = defaultRetries(retries, defaultReadTimeouts())
+	err = retry(
+		fmt.Sprintf("removing VM %s", id),
+		o.logger,
+		retries,
+		func() error {
```
Comment (Gal-Zaidman) [inline client_vm_remove.go:13]: Not critical since it is very small but same as other comments...
I hate that everytime we relay on the parameters (basically on each function that on the oVirtClient struct) we have this anonymous function/internal+wrapping func

File: client_nic_get.go [hunk 9]
```diff
@@ -0,0 +1,41 @@
+package ovirtclient
+
+import (
+	"fmt"
+)
+
+func (o *oVirtClient) GetNIC(vmid string, id string, retries ...RetryStrategy) (result NIC, err error) {
```
Comment (Gal-Zaidman) [inline client_nic_get.go:7]: Why this is not generated like the rest of the get/list?

File: client_nic_list.go [hunk 10]
```diff
@@ -0,0 +1,33 @@
+package ovirtclient
+
+import (
+	"fmt"
+)
+
+func (o *oVirtClient) ListNICs(vmid string, retries ...RetryStrategy) (result []NIC, err error) {
```
Comment (Gal-Zaidman) [inline client_nic_list.go:7]: same as the get comment

File: client_nic.go [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+// NIC represents a network interface.
+type NIC interface {
+	// ID is the identifier for this network interface.
+	ID() string
+	// Name is the user-given name of the network interface.
+	Name() string
+	// VMID is the identified of the VM this NIC is attached to. May be nil if the NIC is not attached.
+	VMID() string
```
Comment (Gal-Zaidman) [inline client_nic.go:26]: Separate PR
I saw that the is a VMs slice on the sdk object, we need to check when it is used and if/how to handle it in the client

File: client_datacenter.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (Gal-Zaidman) [reply client_datacenter.go:19]: I don't think so, the get and list are enough to assume it, the function doc makes sure that the user understands that... its an API to oVirt there are going to be network calls...

**Comment by Gal-Zaidman** [comment]
code looks good to me, just the generation question and I'll approve

File: client_nic.go [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+	// VNICProfileID returns the ID of the VNIC profile in use by the NIC.
+	VNICProfileID() string
+
+	// GetVM fetches an up to date copy of the virtual machine this VM is attached to. This involves an API call and
```
Comment (eslutsky) [inline client_nic.go:30]: looks like typo in the description: `VM is attached to` = `NIC is attached to`

File: client_datacenter.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (janosdebugs) [reply client_datacenter.go:19]: All right, then let's keep it this way.

File: client_nic_get.go (hunk 9 above)
Comment (janosdebugs) [reply client_nic_get.go:7]: After having gone down a rabbit hole, here's the answer: The NIC is a subordinate object to the VM and can only be fetched using the VM ID. We don't have a code generation function for these kinds of objects. I was attempting to correct for that but I can already see that this will be out of scope for this PR.

File: client_nic_list.go (hunk 10 above)
Comment (janosdebugs) [reply client_nic_list.go:7]: See above.

File: errors.go [hunk 13]
```diff
@@ -167,6 +167,10 @@ func (e *engineError) CanAutoRetry() bool {
 	return e.code.CanAutoRetry()
 }
 
+func newFieldNotFound(object string, field string) error {
```
Comment (RomanBednar) [inline errors.go:170]: This error is not used in this commit and thus missing context making a purpose of this error a bit unclear.

File: client_datacenter.go [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+	}
+
+	return &datacenter{
+		client: client,
+		id:     id,
+		name:   name,
+	}, nil
+}
+
+type datacenter struct {
```
Comment (RomanBednar) [inline client_datacenter.go:42]: Inconsistent indentation with other files. For example client_disk.go has different spacing: https://github.com/oVirt/go-ovirt-client/blob/main/client_disk.go#L231

File: client_datacenter_list_clusters.go [hunk 15]
```diff
@@ -0,0 +1,39 @@
+package ovirtclient
```
Comment (RomanBednar) [inline client_datacenter_list_clusters.go:1]: This file is not generated but lives alongside the generated ones. Is there a plan to move generated code into a separate directory?

It would make it more obvious which code is generated from a template and which not, also the versioning would be less messy.

File: client_network.go (hunk 5 above)
Comment (RomanBednar) [reply client_network.go:7]: I noticed there can be up to 3 different names for one thing - ID/name/object - with just different upper/lower chars.

I think this will get very messy over time. Having some kind of mapping or translation at one place would help. But there's not much to be done on our side as this is determined by ovirt SDK?

File: newmock.go [hunk 16]
```diff
@@ -14,9 +14,10 @@ func NewMock() MockClient {
 	testStorageDomain := generateTestStorageDomain()
 	testDatacenter := generateTestDatacenter(testCluster)
 	testNetwork := generateTestNetwork(testDatacenter)
+	testVNICProfile := generateTestVNICProfile(testNetwork)
 	blankTemplate := &template{
-		id:          BlankTemplateID,
-		name:        "Blank",
+		id: BlankTemplateID,
```
Comment (RomanBednar) [inline newmock.go]: Not related - formatting changes in this file should be in a separate commit.

File: test_helper.go [hunk 17]
```diff
@@ -235,10 +280,15 @@ func verifyTestStorageDomainID(client Client, storageDomainID string) error {
 
 type testHelper struct {
 	client          Client
+	rand            *rand.Rand
```
Comment (RomanBednar) [inline test_helper.go:295]: Why move the line? Just to group string type?

File: client_nic.go [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+		id,
+		name,
+		vmid,
+		vnicProfileID,
+	}, nil
+}
+
+type nic struct {
+	client Client
+
```
Comment (RomanBednar) [inline client_nic.go:75]: Extra line and wrong spacing.

File: client_nic_create.go [hunk 19]
```diff
@@ -0,0 +1,60 @@
+package ovirtclient
+
+import (
+	"fmt"
+
```
Comment (RomanBednar) [inline client_nic_create.go:5]: Extra line.

File: client_vm.go [hunk 20]
```diff
@@ -33,6 +33,13 @@ type VM interface {
 
 	// Remove removes the current VM. This involves an API call and may be slow.
 	Remove(retries ...RetryStrategy) error
+
```
Comment (RomanBednar) [inline client_vm.go:44]: Extra line.

File: newmock.go [hunk 21]
```diff
@@ -16,8 +16,8 @@ func NewMock() MockClient {
 	testNetwork := generateTestNetwork(testDatacenter)
 	testVNICProfile := generateTestVNICProfile(testNetwork)
 	blankTemplate := &template{
-		id: BlankTemplateID,
-		name: "Blank",
+		id:          BlankTemplateID,
```
Comment (RomanBednar) [inline newmock.go:19]: Formatting changes should be in separate commit.

File: README.md [hunk 22]
```diff
@@ -118,6 +118,7 @@ func TestSomething(t *testing.T) {
         os.Getenv("OVIRT_CLUSTER_ID"),
         os.Getenv("OVIRT_BLANK_TEMPLATE_ID"),
         os.Getenv("OVIRT_STORAGE_DOMAIN_ID"),
+		os.Getenv("OVIRT_VNIC_PROFILE_ID"),
```
Comment (RomanBednar) [inline README.md]: Wrong indentation?

Comment (janosdebugs) [reply README.md]: Resolved in a later commit.

File: newmock.go (hunk 21 above)
Comment (janosdebugs) [reply newmock.go:19]: All files are formatted using gofmt, otherwise this PR would not be merged. In Go, formatting changes are acceptable so long as they are done with gofmt.

File: client_vm.go (hunk 20 above)
Comment (janosdebugs) [reply client_vm.go:44]: This line has been added to separate the NIC functions from the VM-internal functions (logical separation).

File: client_nic_create.go (hunk 19 above)
Comment (janosdebugs) [reply client_nic_create.go:5]: This line separates the Go-internal packages (`fmt`) from the imported third party packages (`ovirtsdk`). This is done automatically by the IDE and is enforced by our linter.

File: client_nic.go (hunk 18 above)
Comment (janosdebugs) [reply client_nic.go:75]: The extra line has been added to separate the client (utility) from the data storage. The spacing is done by gofmt and is correct.

File: client_storagedomain.go [hunk 23]
```diff
@@ -94,6 +96,8 @@ func convertSDKStorageDomain(sdkStorageDomain *ovirtsdk4.StorageDomain) (Storage
 }
 
 type storageDomain struct {
+	client Client
```
[the remark this answers is not in the document]
Comment (janosdebugs) [reply client_storagedomain.go:99]: Indentation is correct in all cases, otherwise it wouldn't pass CI. Newlines after client are added to create a logical separation between the client and data structure components.

File: client_storagedomain.go [hunk 24]
```diff
@@ -85,6 +85,8 @@ func convertSDKStorageDomain(sdkStorageDomain *ovirtsdk4.StorageDomain) (Storage
 	}
 
 	return &storageDomain{
+		client: client,
```
[the remark this answers is not in the document]
Comment (janosdebugs) [reply client_storagedomain.go:88]: Spacing is done by gofmt, otherwise it wouldn't pass CI.

File: scripts/rest.go [hunk 25]
```diff
@@ -102,6 +110,12 @@ func setupFlags(name *string, id *string, tplDir *string, targetDir *string, wat
 		"",
 		"Pass an identifier used in the SDK. Must be capitalized. E.g. \"StorageDomain\". Required.",
 	)
+	flag.StringVar(
+		object,
+		"o",
+		"",
+		"Pass an identifier used in the client. Defaults to the same value as -i. Must be capitalized. E.g. \"StorageDomain\".",
```
[the remark this answers is not in the document]
Comment (janosdebugs) [reply scripts/rest.go:131]: `-o` has a default value, not `-i`. `-i` must be explicitly provided. Will add an `os.Exit` if -i is unset.

File: errors.go (hunk 13 above)
Comment (janosdebugs) [reply errors.go:170]: This is a shorthand for creating an `EFieldMissing` error. See that error for a description.

File: client_datacenter_list_clusters.go (hunk 15 above)
Comment (janosdebugs) [reply client_datacenter_list_clusters.go:1]: No, we cannot move generated files into a separate directory because that would put them into a separate package. Unfortunately, that is not a thing we can do.

File: newmock.go (hunk 16 above)
Comment (janosdebugs) [reply newmock.go]: This is done by gofmt and is legit.

File: test_helper.go (hunk 17 above)
Comment (janosdebugs) [reply test_helper.go:295]: Yes, make it a bit easier to read. Down the line (separate PR) we will reorder all structs so that struct fields align and use less memory.

File: client_nic_get.go (hunk 9 above)
Comment (Gal-Zaidman) [reply client_nic_get.go:7]: ok makes sense, if this will be a repeated pattern then we should probably have generation for that case as well but for now, it is ok

**Review by Gal-Zaidman — APPROVED** [review]
Changes looks good to me
